<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Alan Selewa" />


<title>Assessing Seurat Integration</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Integration Project</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Assessing Seurat Integration</h1>
<h4 class="author">Alan Selewa</h4>

</div>


<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>I obtained paired scRNA-seq and scATAC-seq from 10X Genomics for paired <a href="https://support.10xgenomics.com/single-cell-multiome-atac-gex/datasets/1.0.0/pbmc_unsorted_10k">PBMC cells</a>. I performed unsupervised clustering on the scRNA-seq data using marker genes outlined in the Seurat 3k PBMC tutorial. The cell-types are shown below.</p>
<pre class="r"><code>s.rna &lt;- readRDS(&#39;data/Seurat/rna_seurat.rds&#39;)
s.atac &lt;- readRDS(&#39;data/Seurat/atac_seurat.rds&#39;)
anchors &lt;- readRDS(&#39;data/anchors/anchors_kanchor_5.rds&#39;)</code></pre>
<pre class="r"><code>p1 &lt;- DimPlot(s.rna, label=T, group.by=&#39;clusters&#39;) + NoLegend() + ggtitle(&#39;scRNA-seq&#39;)
p2 &lt;- DimPlot(s.atac, label=T, group.by=&#39;clusters&#39;) + NoLegend() + ggtitle(&#39;scATAC-seq&#39;)
p1 + p2</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="evaluating-anchors" class="section level2">
<h2>Evaluating anchors</h2>
<p>Using default parameters, I ran <code>FindIntegrationAnchors</code> from Seurat v3 on the scRNA-seq and scATAC-seq data above, and obtained 5658 anchor pairs. Below we evaluate how accurate these anchors are.</p>
<pre class="r"><code>cells1 &lt;- anchors@anchors[,&quot;cell1&quot;]
cells2 &lt;- anchors@anchors[,&quot;cell2&quot;]</code></pre>
<pre class="r"><code>num_cells &lt;- c(length(unique(cells1)), length(unique(cells2)))/length(Cells(s.rna))
names(num_cells) &lt;- c(&#39;RNA&#39;, &#39;ATAC&#39;)
barplot(num_cells, ylab=&#39;Proportion of Cells&#39;,xlab=&#39;Assay&#39;, main=&#39;Contribution to Anchors&#39;)</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<pre class="r"><code>freq_cells1 &lt;- table(cells1)
hist_cells1 &lt;- rep(0, max(freq_cells1))
for(i in 1:max(freq_cells1)){
  hist_cells1[i] &lt;- sum(freq_cells1==i)
}
barplot(hist_cells1, names.arg = 1:max(freq_cells1), ylab=&#39;Number of Cells&#39;, xlab=&#39;Frequency&#39;, main=&#39;Frequency of cells used in Anchors (RNA)&#39;)</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<pre class="r"><code>same_cell &lt;- c(sum(cells1 == cells2), sum(cells1 != cells2))
names(same_cell) &lt;- c(&#39;same cell&#39;, &#39;different cell&#39;)
barplot(same_cell, ylab=&#39;Anchors&#39;,xlab=&#39;Anchor Pairs&#39;, main=&#39;Anchor Pair Correspondence&#39;)</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<pre class="r"><code>cells1_clust &lt;- as.character(s.rna$clusters[cells1])
cells2_clust &lt;- as.character(s.atac$clusters[cells2])
cell_type_agree &lt;- c(sum(cells1_clust != cells2_clust), sum(cells1_clust == cells2_clust))
names(cell_type_agree) &lt;- c(&#39;Disagree&#39;, &#39;Agree&#39;)

scores_agree &lt;- anchors@anchors[cells1_clust == cells2_clust, &quot;score&quot;]
scores_disagree &lt;- anchors@anchors[cells1_clust != cells2_clust, &quot;score&quot;]

par(mfrow=c(1,2))
barplot(cell_type_agree, ylab=&#39;Anchors&#39;,xlab=&#39;Cell Type Assignment&#39;, main=&#39;Anchors Cluster Agreement&#39;)
boxplot(scores_disagree, scores_agree, names = c(&#39;Disagree&#39;,&#39;Agree&#39;), ylab=&#39;Anchor Score&#39;, main=&#39;Anchor Scores&#39;)</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
</div>
<div id="label-transfer" class="section level2">
<h2>Label Transfer</h2>
<p>Having obtained anchors, we can perform label transfer from RNA to ATAC. Below I show the the ATAC UMAP with labels transferred from RNA.</p>
<pre class="r"><code>s.atac &lt;- suppressMessages(transfer_data(ATAC = s.atac, data = s.rna$clusters, anchors = anchors))
p1 &lt;- DimPlot(s.rna, label=T) + NoLegend() + ggtitle(&#39;scRNA-seq&#39;)
p2 &lt;- DimPlot(s.atac, label=T, group.by=&#39;RNA_IDENT&#39;) + NoLegend() + ggtitle(&#39;scATAC-seq&#39;)
p1 + p2</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>Because we know the actual cell-type label of ATAC cells (based on their paired RNA data), we can ask if the transferred label agrees with the actual label.</p>
<pre class="r"><code>cell_type_agree &lt;- c(sum(s.atac$RNA_IDENT != s.atac$clusters), sum(s.atac$RNA_IDENT == s.atac$clusters))
names(cell_type_agree) &lt;- c(&#39;Disagree&#39;, &#39;Agree&#39;)

scores_agree &lt;- s.atac$PREDICTED_SCORE_MAX[s.atac$RNA_IDENT != s.atac$clusters]
scores_disagree &lt;- s.atac$PREDICTED_SCORE_MAX[s.atac$RNA_IDENT == s.atac$clusters]

par(mfrow=c(1,2))
barplot(cell_type_agree, ylab=&#39;ATAC Cells&#39;,xlab=&#39;Transferred Label Status&#39;, main=&#39;ATAC Label Transfer&#39;)
boxplot(scores_agree, scores_disagree, names = c(&#39;Disagree&#39;,&#39;Agree&#39;), ylab=&#39;Max Predicted Score&#39;, xlab=&#39;Transferred Label Status&#39;, main=&#39;ATAC Label Transfer Max Score&#39;)</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<div id="benchmarking-label-transfer" class="section level3">
<h3>Benchmarking label transfer</h3>
<p>Here we look at the performance of label transfer under different clustering resolutions. The benchmark procedure is summarized below:</p>
<ol style="list-style-type: decimal">
<li>Generate clusters for each cell based on RNA alone</li>
<li>Transfer label from RNA to ATAC in an unsupervised fashion</li>
<li>Compare transfered label with the label generated in 1.</li>
</ol>
<p>We get this accuracy at various levels of clustering resolution.</p>
<p>Furthermore, we tune 2 parameters:</p>
<ul>
<li>k.anchor: the size of the neighborhood for finding mutual nearest neighbors.</li>
<li>k.filter: filters anchors that are NOT within <code>k.filter</code> nearest neighbors of eachother in high dimensional space (gene space).</li>
</ul>
<p>The transfer label performs worse as the number of clusters is increased. Here the range of resolution implies 5 to 18 clusters.</p>
<pre class="r"><code>lab.acc.list &lt;- readRDS(&#39;data/transfer_label_benchmarks/label_transfer_accuracy.rds&#39;)
res &lt;- readRDS(&#39;data/transfer_label_benchmarks/resolutions.rds&#39;)

lab.acc.smooth &lt;- lapply(lab.acc.list, FUN = function(x){zoo::rollmean(x = x, k=3)})
res &lt;- res[1:length(lab.acc.smooth$`1`)]

benchmark.df &lt;- data.frame(lab.acc.smooth)
colnames(benchmark.df) &lt;- as.character(c(1,10,15,20,5))
benchmark.df &lt;- suppressMessages(reshape2::melt(benchmark.df))
colnames(benchmark.df) &lt;- c(&quot;k.anchor&quot;,&quot;TransferLabelAcurracy&quot;)
benchmark.df$res &lt;- rep(res, length(lab.acc.smooth))</code></pre>
<pre class="r"><code>p1 &lt;- ggplot(benchmark.df, aes(x=res, y=TransferLabelAcurracy, color=k.anchor)) + geom_point() + geom_line() + ggClean() + xlab(&#39;Resolution&#39;) + ylab(&#39;Transfer Label Accuracy&#39;) + ggtitle(&#39;kanchor benchmark&#39;)</code></pre>
<pre class="r"><code>lab.acc.list &lt;- readRDS(&#39;data/transfer_label_benchmarks/label_transfer_accuracy_kfilter.rds&#39;)
res &lt;- readRDS(&#39;data/transfer_label_benchmarks/resolutions.rds&#39;)

lab.acc.smooth &lt;- lapply(lab.acc.list, FUN = function(x){zoo::rollmean(x = x, k=3)})
res &lt;- res[1:length(lab.acc.smooth$`10`)]

benchmark.df &lt;- data.frame(lab.acc.smooth)
colnames(benchmark.df) &lt;- as.character(c(10,100,200,500,Inf))
benchmark.df &lt;- suppressMessages(reshape2::melt(benchmark.df))
colnames(benchmark.df) &lt;- c(&quot;k.filter&quot;,&quot;TransferLabelAcurracy&quot;)
benchmark.df$res &lt;- rep(res, length(lab.acc.smooth))</code></pre>
<pre class="r"><code>p2 &lt;- ggplot(benchmark.df, aes(x=res, y=TransferLabelAcurracy, color=k.filter)) + geom_point() + geom_line() + ggClean() + xlab(&#39;Resolution&#39;) + ylab(&#39;Transfer Label Accuracy&#39;) + ggtitle(&#39;kfilter benchmark&#39;)</code></pre>
<pre class="r"><code>p1 + p2</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-20-1.png" width="960" /></p>
<p>Based on the above results, it seems that the metric is fairly robust to the <code>k.anchor</code> parameter. Furthermore, it seems that filtering anchors according to high dimension generally not a good idea because turning off the <code>k.filter</code> by setting it to <code>Inf</code> results in using all anchors and achieves higher accuracy.</p>
</div>
</div>
<div id="rna-transfer" class="section level2">
<h2>RNA Transfer</h2>
<p>We can also transfer gene expression from RNA to ATAC. But how good is the transfer of RNA data? Here we explore this by looking at peak-gene correlations in close detail. In particular, we grab all genes within 100kb of each peak (~200k pairs) and then find their correlations using pseudo-cells. A pseudo-cell here contains aggregated read counts for RNA and ATAC-seq across 28 cells.</p>
<div id="simple-correlations-before-vs-after-transfer" class="section level3">
<h3>Simple correlations before vs after transfer</h3>
<p>Lets transfer some data and see how the cells and genes correlate with the original RNA data.</p>
<pre class="r"><code>og.data &lt;- as.matrix(s.rna@assays$RNA@data)
imputed.rna &lt;- suppressMessages(TransferData(anchorset = anchors, refdata = og.data, weight.reduction = &#39;cca&#39;))</code></pre>
<pre class="r"><code>imputed.rna.mat &lt;- as.matrix(imputed.rna@data)
topN &lt;- order(rowSums(og.data), decreasing = T)[1:100]
topN.imputed &lt;- imputed.rna.mat[topN,]
topN.og &lt;- og.data[topN,]
gene.cor &lt;- rep(0, length(topN))
for(i in 1:length(topN)){
  x &lt;- topN.imputed[i,]
  y &lt;- topN.og[i,]
  gene.cor[i] &lt;- cor(x, y, method = &#39;pearson&#39;)
}</code></pre>
<pre class="r"><code>n &lt;- dim(imputed.rna.mat)[2]
cell.cor &lt;- rep(0, n)
for(i in 1:n){
  bol &lt;- imputed.rna.mat[,i] &gt; 0
  x &lt;- imputed.rna.mat[,i][bol]
  y &lt;- og.data[,i][bol]
  cell.cor[i] &lt;- cor(x, y, method = &#39;pearson&#39;)
}</code></pre>
<pre class="r"><code>par(mfrow=c(1,2))
hist(cell.cor, xlab=&#39;Pearson&#39;, main=&#39;Cells&#39;)
hist(gene.cor, xlab=&#39;Pearson&#39;, main=&#39;Top 100 Genes&#39;)</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>
</div>
<div id="peak-gene-correlations-in-pseudo-cells" class="section level3">
<h3>Peak-gene correlations in pseudo-cells</h3>
<p>Here, we find all genes within 100kb of all peaks in the scATAC-seq data (~62k peaks). We will use aggregated RNA and ATAC-seq (pseudo-cells) to find robust correlations.</p>
<pre class="r"><code>satac_agg &lt;- readRDS(&#39;data/Seurat/agg_atac_seurat.rds&#39;)
srna_agg &lt;- readRDS(&#39;data/Seurat/agg_rna_seurat.rds&#39;)</code></pre>
<p>Here each each dot represents ~28 cells. The pseudo-cells are labeled by the labels of their RNA counterpart. As we can see, the pseudo-cells cluster together.</p>
<pre class="r"><code>DimPlot(satac_agg, group.by=&#39;clusters&#39;, label=T) + LegendOff() + ggtitle(&#39;Pseuo-cell ATAC-seq&#39;)</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>
<p>Lets find every gene within 100kb of each peak.</p>
<pre class="r"><code>annotations &lt;- suppressWarnings(suppressMessages(GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, verbose = F)))
ensembldb::seqlevelsStyle(annotations) &lt;- &quot;UCSC&quot;
exon_annotations &lt;- annotations[annotations$type==&quot;exon&quot;,]

DA_peaks_GR &lt;- granges(satac_agg)
DA_peaks_genes &lt;- suppressWarnings(plyranges::join_overlap_left(x = DA_peaks_GR, y = exon_annotations, maxgap=100000, minoverlap=0))
peakies &lt;- Signac::GRangesToString(DA_peaks_genes)
genes &lt;- DA_peaks_genes$gene_name
comb &lt;- paste0(peakies, &#39;-&#39;, genes)
peakies &lt;- peakies[!duplicated(comb)]
genes &lt;- genes[!duplicated(comb)]

enhancers_genes_df &lt;- cbind(peakies, genes)
enhancers_genes_df &lt;- enhancers_genes_df[!is.na(enhancers_genes_df[,&quot;genes&quot;]),]</code></pre>
<p>Pull aggregated pseudo-cell data.</p>
<pre class="r"><code>agg.atac.mat &lt;- as.matrix(satac_agg@assays$peaks@data)
agg.rna.mat &lt;- as.matrix(srna_agg@assays$RNA@data)

enhancers_genes_df &lt;- enhancers_genes_df[enhancers_genes_df[,&quot;genes&quot;] %in% rownames(agg.rna.mat),]
DA.atac.mat &lt;- agg.atac.mat[enhancers_genes_df[,&quot;peakies&quot;],]
genes.DA.rna.mat &lt;- agg.rna.mat[enhancers_genes_df[,&quot;genes&quot;],]</code></pre>
<p>Get correlations between peaks and true RNA-seq data across all cells</p>
<pre class="r"><code>row_by_row_cor &lt;- function(A, B){
  p1 &lt;- dim(A)[1]
  p2 &lt;- dim(B)[1]
  stopifnot(p1 == p2)
  rr_cors &lt;- rep(0, p1)
  for(i in 1:p1){
    x &lt;- A[i,]
    y &lt;- B[i,]
    if(stats::sd(x) &gt; 0 &amp; stats::sd(y) &gt; 0){
      rr_cors[i] &lt;- cor(x, y)
    }
    else{
      rr_cors[i] &lt;- NA
    }
  }
  return(rr_cors)
}</code></pre>
<pre class="r"><code>gene_enhancer_cor &lt;- row_by_row_cor(A = DA.atac.mat, B = genes.DA.rna.mat)</code></pre>
<p>Transfer RNA counts to ATAC-seq cells</p>
<pre class="r"><code>og.data &lt;- as.matrix(s.rna@assays$RNA@counts)
imputed.rna &lt;- suppressMessages(TransferData(anchorset = anchors, refdata = og.data, weight.reduction = &#39;cca&#39;))</code></pre>
<p>Aggregate the transferred RNA data, so we can compare to the original aggregated data.</p>
<pre class="r"><code>aggregate.counts &lt;- list()
imputed.counts &lt;- as.matrix(imputed.rna@data)
for(i in unique(s.rna$seurat_clusters)){
  sub.counts &lt;- imputed.counts[,s.rna$seurat_clusters == i]
  aggregate.counts[[i]] &lt;- rowSums(sub.counts)
}
aggregate.counts &lt;- as.data.frame(aggregate.counts)
colnames(aggregate.counts) &lt;- 1:ncol(aggregate.counts)

imputed_srna_agg &lt;- CreateSeuratObject(counts = aggregate.counts)
imputed_srna_agg &lt;- NormalizeData(imputed_srna_agg)
imputed_srna_agg_mat &lt;- as.matrix(imputed_srna_agg@assays$RNA@data)</code></pre>
<p>Get correlations between peaks and imputed/transferred RNA-seq data across all cells</p>
<pre class="r"><code>genes.DA.rna.imputed.mat &lt;- imputed_srna_agg_mat[enhancers_genes_df[,&quot;genes&quot;],]

gene_enhancer_cor_imputed &lt;- row_by_row_cor(A = DA.atac.mat, B = genes.DA.rna.imputed.mat)</code></pre>
<p>Correlations before vs after imputation</p>
<pre class="r"><code>plot.df &lt;- data.frame(imputed_cor = gene_enhancer_cor_imputed, true_cor=gene_enhancer_cor)
plot.df &lt;- plot.df %&gt;% drop_na()
get_density &lt;- function(x, y, ...) {
  dens &lt;- MASS::kde2d(x, y, ...)
  ix &lt;- findInterval(x, dens$x)
  iy &lt;- findInterval(y, dens$y)
  ii &lt;- cbind(ix, iy)
  return(dens$z[ii])
}
plot.df$density &lt;- get_density(plot.df$imputed_cor, plot.df$true_cor)

ggplot(plot.df, aes(x=imputed_cor, y=true_cor, color=density)) + geom_point() + ggtitle(&#39;Gene-Peak Correlation with Transfered RNA&#39;) + xlab(&#39;Pearson (imputed)&#39;) + ylab(&#39;Pearson (original)&#39;) + geom_abline(slope = 1, color=&quot;red&quot;, linetype=&quot;dashed&quot;) + ggClean() + scale_color_viridis()</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-34-1.png" width="672" /></p>
</div>
<div id="precision-recall-of-gene-peaks-detected" class="section level3">
<h3>Precision-recall of gene-peaks detected</h3>
<p>Here we plot the recall vs precision at various cutoffs of the Pearson correlation coefficient. The recall/precision are computed w.r.t the gene-peak correlations in the original multiome data (with pseudo-cells)</p>
<pre class="r"><code>cors &lt;- seq(0.1, 0.9, by=0.05)
recall &lt;- rep(0, length(cors))
precision &lt;- rep(0, length(cors))
peak_gene_str &lt;- paste0(enhancers_genes_df[,&quot;peakies&quot;],&#39;-&#39;,enhancers_genes_df[,&quot;genes&quot;])

for(i in 1:length(cors)){
  
  ground_truth_pairs &lt;- peak_gene_str[which(abs(gene_enhancer_cor) &gt; cors[i])]
  inferred_pairs &lt;- peak_gene_str[which(abs(gene_enhancer_cor_imputed) &gt; cors[i])]

  recall[i] &lt;- length(intersect(inferred_pairs,ground_truth_pairs))/length(ground_truth_pairs)
  precision[i] &lt;- length(intersect(inferred_pairs,ground_truth_pairs))/length(inferred_pairs)
  
}</code></pre>
<pre class="r"><code>re.pre.df &lt;- data.frame(recall=recall, precision=precision, pearson=cors)
ggplot(re.pre.df, aes(x=precision, y = recall, color=pearson)) + geom_point() + geom_line() + ggClean()</code></pre>
<p><img src="integration_workflow_files/figure-html/unnamed-chunk-36-1.png" width="672" /></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
